content/recipes/swagger.md

```diff
@@ -348,52 +412,219 @@ Let's open the browser and verify the generated `Cat` model:
+})
 uploadFile(@UploadedFile() file) {}
 ```
 
 #### Decorators
 
-All of the available OpenAPI decorators has an `Api` prefix to be clearly distinguishable from the core decorators. Below is a full list of the exported decorators with a defined use-level (where might be applied).
-
-|                               |                     |
-| ----------------------------- | ------------------- |
-| `@ApiOperation()`             | Method              |
-| `@ApiResponse()`              | Method / Controller |
-| `@ApiProduces()`              | Method / Controller |
-| `@ApiConsumes()`              | Method / Controller |
-| `@ApiBearerAuth()`            | Method / Controller |
-| `@ApiOAuth2Auth()`            | Method / Controller |
-| `@ApiImplicitBody()`          | Method              |
-| `@ApiImplicitParam()`         | Method              |
-| `@ApiImplicitQuery()`         | Method              |
-| `@ApiImplicitHeader()`        | Method              |
-| `@ApiImplicitFile()`          | Method              |
-| `@ApiExcludeEndpoint()`       | Method              |
-| `@ApiUseTags()`               | Method / Controller |
-| `@ApiModelProperty()`         | Model               |
-| `@ApiModelPropertyOptional()` | Model               |
+All of the available OpenAPI decorators has an `Api` prefix to be easily distinguishable from the core decorators. Below is a full list of the exported decorators with a defined use-level (where can be applied).
+
+|                          |                     |
+| ------------------------ | ------------------- |
+| `@ApiOperation()`        | Method              |
+| `@ApiResponse()`         | Method / Controller |
+| `@ApiProduces()`         | Method / Controller |
+| `@ApiConsumes()`         | Method / Controller |
+| `@ApiBearerAuth()`       | Method / Controller |
+| `@ApiOAuth2()`           | Method / Controller |
+| `@ApiBasicAuth()`        | Method / Controller |
+| `@ApiSecurity()`         | Method / Controller |
+| `@ApiExtraModels()`      | Method / Controller |
+| `@ApiBody()`             | Method              |
+| `@ApiParam()`            | Method              |
+| `@ApiQuery()`            | Method              |
+| `@ApiHeader()`           | Method / Controller |
+| `@ApiExcludeEndpoint()`  | Method              |
+| `@ApiTags()`             | Method / Controller |
+| `@ApiProperty()`         | Model               |
+| `@ApiPropertyOptional()` | Model               |
+| `@ApiHideProperty()`     | Model               |
+
+#### Plugin
+
+TypeScript metadata reflaction system has several limitations, which make it impossible to, for instance, determine what properties a class consist of or recognize whether a given property is optional or required. However, some of these constraints can be addressed at the compilation time. Hence, we created a plugin that enhances TypeScript compilation process to reduce the boilerplate code.
+
+Swagger plugin will automatically:
+
```

Comment:
The Swagger plugin will automatically:

- annotate all DTO properties with `@ApiProperty` unless `@ApiHideProperty` is used
- set the `required` property depending on the question mark (e.g. `name?: string` will set `required: false`)
- set the `type` or `enum` property depending on the type (supports arrays as well)
- set the `default` property based on the assigned default value
- set several validation rules based on `class-validator` decorators (if `classValidatorShim` set to `true`)
- add a response decorator to every endpoint with a proper status and `type` (response model)

In order to enable the plugin, simply open `nest-cli.json` (if you use [Nest CLI](/cli/overview)) and add the following `plugins` configuration: